[DCP] Remove all-gather of state dict keys

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145998

The original `_all_gather_keys` call was for a safety check, but could be costly as things scale, and it blocks CPU.

Instead, we make it clear in the documentation that the `state_dict` passed to the `load` API should have same set of keys, otherwise the API may hang. 

In addition, we move the check to a utility function: `utils.assert_same_keys`. User uncertain about state dict unity can optionally call this API to check.

Resolves #145965 (as a workaround).

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @ekr0

kwen2501 committed:
[DCP] Remove all-gather of state dict keys; fix #145965

[ghstack-poisoned]

kwen2501 committed:
Update on "[DCP] Remove all-gather of state dict keys"


The original `_all_gather_keys` call was for a safety check, but could be costly as things scale, and it blocks CPU.

Instead, we make it clear in the documentation that the `state_dict` passed to the `load` API should have same set of keys, otherwise the API may hang. 

Resolves #145965 (as a workaround).

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz pradeepfn ekr0

[ghstack-poisoned]

Skylion007 commented on torch/distributed/checkpoint/utils.py:
```diff
@@ -42,13 +42,32 @@ def _get_failure_dict(
+
+
+def assert_same_keys(
+    state_dict: dict[Any, Any], process_group: Optional[dist.ProcessGroup] = None
```
Use TypeVar for keys in dict here too

kwen2501 committed:
Update on "[DCP] Remove all-gather of state dict keys"


The original `_all_gather_keys` call was for a safety check, but could be costly as things scale, and it blocks CPU.

Instead, we make it clear in the documentation that the `state_dict` passed to the `load` API should have same set of keys, otherwise the API may hang. 

In addition, we move the check to a utility function: `utils.assert_same_keys`. User uncertain about state dict unity can optionally call this API to check.

Resolves #145965 (as a workaround).

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz pradeepfn ekr0

[ghstack-poisoned]

kwen2501 committed:
Update on "[DCP] Remove all-gather of state dict keys"


The original `_all_gather_keys` call was for a safety check, but could be costly as things scale, and it blocks CPU.

Instead, we make it clear in the documentation that the `state_dict` passed to the `load` API should have same set of keys, otherwise the API may hang. 

In addition, we move the check to a utility function: `utils.assert_same_keys`. User uncertain about state dict unity can optionally call this API to check.

Resolves #145965 (as a workaround).

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz pradeepfn ekr0

[ghstack-poisoned]

mhorowitz commented on torch/distributed/checkpoint/utils.py:
```diff
@@ -41,14 +41,33 @@ def _get_failure_dict(
+    my_keys = set(state_dict.keys())
+    diff = all_keys - my_keys
+    if len(diff) > 0:
+        raise AssertionError(f"Keys mismatch between ranks, difference: {diff}")
```
This actually detects something slightly different, which is if other ranks include a key not included in this rank.  If two other ranks have different keys, but the union is the same as this one, this will not raise, but one of the others will.  If this rank has an extra key, this will not raise, but other ranks will.
```suggestion
        raise AssertionError(f"Key(s) present in other ranks but not this one, difference: {diff}")
```

kwen2501 replied:
sg

mhorowitz commented on torch/distributed/checkpoint/utils.py:
```diff
@@ -9,7 +9,7 @@
 from functools import wraps
 from pstats import Stats
-from typing import Any, Callable, cast, Optional, TypeVar, Union
+from typing import Any, Callable, cast, Optional, Set, TypeVar, Union
```
don't change. see below

mhorowitz commented on torch/distributed/checkpoint/state_dict_loader.py:
```diff
@@ -59,7 +59,11 @@ def load(
+
+    Each rank must have the same keys in their ``state_dict`` provided to this
+    API. Mismatched keys may result in hangs or errors. If unsure, you can use
+    the ``assert_same_keys`` API to check (but may induce communication costs).
```
```suggestion
    the ``_assert_same_keys`` API to check (but may incur communication costs).
```

mhorowitz reviewed (approved):
Some nits, but I think this is ok to land, once you've fixed those, and the failing doc test (which I think will disappear if you rename the function as suggested).

kwen2501 commented on torch/distributed/checkpoint/utils.py:
```diff
@@ -41,14 +41,33 @@ def _get_failure_dict(
+    join. It will also induce cross-rank communication and block CPU.
+    """
+
+    if dist.get_world_size(process_group) == 1:
```
The `get_world_size` API should support `None`.

mhorowitz replied:
Empirically, it does not:
```
$ python
Python 3.12.7 | packaged by Anaconda, Inc. | (main, Oct  4 2024, 13:27:36) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.distributed.get_world_size(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/users/mhorowitz/pytorch/torch/distributed/distributed_c10d.py", line 2329, in get_world_size
    return _get_group_size(group)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/mhorowitz/pytorch/torch/distributed/distributed_c10d.py", line 1096, in _get_group_size
    default_pg = _get_default_group()
                 ^^^^^^^^^^^^^^^^^^^^
  File "/data/users/mhorowitz/pytorch/torch/distributed/distributed_c10d.py", line 1302, in _get_default_group
    raise ValueError(
ValueError: Default process group has not been initialized, please make sure to call init_process_group.`
```

kwen2501 commented on torch/distributed/checkpoint/utils.py:
```diff
@@ -41,14 +41,33 @@ def _get_failure_dict(
+    return set(itertools.chain.from_iterable(gathered_keys))
+
+
+def assert_same_keys(
```
sg

kwen2501 committed:
Update on "[DCP] Remove all-gather of state dict keys"


The original `_all_gather_keys` call was for a safety check, but could be costly as things scale, and it blocks CPU.

Instead, we make it clear in the documentation that the `state_dict` passed to the `load` API should have same set of keys, otherwise the API may hang. 

In addition, we move the check to a utility function: `utils.assert_same_keys`. User uncertain about state dict unity can optionally call this API to check.

Resolves #145965 (as a workaround).

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz pradeepfn ekr0

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge

kwen2501 committed:
Update on "[DCP] Remove all-gather of state dict keys"


The original `_all_gather_keys` call was for a safety check, but could be costly as things scale, and it blocks CPU.

Instead, we make it clear in the documentation that the `state_dict` passed to the `load` API should have same set of keys, otherwise the API may hang. 

In addition, we move the check to a utility function: `utils.assert_same_keys`. User uncertain about state dict unity can optionally call this API to check.

Resolves #145965 (as a workaround).

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz pradeepfn ekr0

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge
